py/server/deephaven/table_listener.py

```diff
@@ -72,208 +257,164 @@ def _do_locked(f: Callable, lock_type="shared") -> None:
+        ...
 
-    if callable(listener):
-        f = listener
-    elif hasattr(listener, "onUpdate"):
-        f = listener.onUpdate
-    else:
-        raise ValueError(
-            "ShiftObliviousListener is neither callable nor has an 'onUpdate' method"
-        )
 
-    return len(signature(f).parameters)
+def listener_wrapper(table: Table):
+    def decorator(listener: Callable):
+        @wraps(listener)
+        def wrapper(update, *args):
+            sig = inspect.signature(listener)
+            n_params = len(sig.parameters)
+            t_update = TableUpdate(table=table, j_table_update=update)
 
+            if n_params == 2:
+                listener(t_update, args[0])
+            else:
+                listener(t_update)
 
-class TableListener(ABC):
-    """An abstract table listener class that should be subclassed by any user Table listener class."""
+        return wrapper
 
-    @abstractmethod
-    def onUpdate(self, *args, **kwargs) -> None:
-        ...
+    return decorator
 
 
-def listen(
-    t: Table,
-    listener: Union[Callable, Type[TableListener]],
-    description: str = None,
-    retain: bool = True,
-    listener_type: str = "auto",
-    start_listening: bool = True,
-    replay_initial: bool = False,
-    lock_type: str = "shared",
-) -> TableListenerHandle:
-    """Listen to table changes.
+def _wrap_listener_func(t: Table, description: str, listener: Callable):
+    n_params = len(signature(listener).parameters)
+    if n_params != 2:
+        raise ValueError("listener must have 2 (update, is_replay) parameters.")
+    return listener_wrapper(table=t)(listener)
 
-    Table change events are processed by 'listener', which can be either
-        (1) a callable (e.g. function) or
-        (2) an object that provides an "onUpdate" method.
-    In either case, the method must have one of the following signatures.
 
-    * (added, removed, modified): shift_oblivious
-    * (isReplay, added, removed, modified): shift_oblivious + replay
-    * (update): shift-aware
-    * (isReplay, update): shift-aware + replay
+def _wrap_listener_obj(t: Table, description: str, listener: TableListener):
+    n_params = len(signature(listener.on_update).parameters)
+    if n_params != 2:
+        raise ValueError(f"The on_update method must have 2 (update, is_replay) parameters.")
+    listener.on_update = listener_wrapper(table=t)(listener.on_update)
+    return listener
 
-    For shift-oblivious listeners, 'added', 'removed', and 'modified' are the indices of the rows which changed.
-    For shift-aware listeners, 'update' is an object that describes the table update.
 
-    Listeners that support replaying the initial table snapshot have an additional parameter, 'isReplay', which is
-    true when replaying the initial snapshot and false during normal updates.
+def listen(t: Table, listener: Union[Callable, TableListener], description: str = None, do_replay: bool = False,
```

Comment:
See my earlier comment about the API confusion I had about if the listener was started or not.
... but you very clearly document the behavior, so I may be an idiot.  What is here may be ok.  I'm leaving my thoughts so that you can think about it.